In [1]:

!pip install torch-geometric -q
!pip install rdkit
import torch
from torch.utils.data import IterableDataset
from torch_geometric.datasets import MoleculeNet
import pandas as pd
import itertools
import sys
from google.colab import drive
import os
#I presume you placed the 'Project_HIV' folder within your main drive folder.
gdrive_path='/content/gdrive/MyDrive/Project_HIV'


drive.mount('/content/gdrive', force_remount=True)

os.chdir(gdrive_path)

print(sorted(os.listdir()))


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.1/63.1 kB 5.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 62.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.2/36.2 MB 57.5 MB/s eta 0:00:00
Mounted at /content/gdrive
['.git', '.gitignore', 'HIV.ipynb', 'README.md', '__pycache__', 'code', 'data_import.py', 'datasets', 'hiv_test.jsonl', 'hiv_train.jsonl', 'hiv_val.jsonl', 'model_with_pe.pth', 'model_without_pe.pth', 'utils.py']


In [2]:
from torch_geometric.datasets import MoleculeNet
class PartialHIVDataset(IterableDataset):
    """
    IterableDataset that loads only a small portion of the HIV dataset.
    """

    def __init__(self, root='/tmp/HIV', max_samples=10):
        """
        Args:
            root: Root directory for dataset storage
            max_samples: Maximum number of samples to load
        """
        self.root = root
        self.max_samples = max_samples
        self._dataset = None

    def _lazy_load_dataset(self):
        """Lazy load the dataset only when iteration begins"""
        if self._dataset is None:
            print(f"Initializing HIV dataset (will only load {self.max_samples} samples)...")
            # Import here to ensure it's available

            self._dataset = MoleculeNet(root=self.root, name='HIV')
            print(f"Dataset ready! Total size: {len(self._dataset)} molecules")
            print(f"But we'll only load {self.max_samples} of them.\n")

    def parse_molecules(self):
        """
        Parse molecules from the dataset, stopping after max_samples.
        """
        self._lazy_load_dataset()

        for i in range(min(self.max_samples, len(self._dataset))):
            data = self._dataset[i]

            # Extract molecule information
            mol_info = {
                'source': f"molecule_{i}",
                'target': data.y.item(),    # HIV activity label
                'num_atoms': data.num_nodes,
                'num_bonds': data.num_edges // 2,
                'smiles': getattr(data, 'smiles', 'N/A')
            }

            yield mol_info

    def __iter__(self):
        """Iterator with worker support"""
        iterator = self.parse_molecules()
        worker_info = torch.utils.data.get_worker_info()

        if worker_info is not None:
            worker_total_num = worker_info.num_workers
            worker_id = worker_info.id
            return itertools.islice(iterator, worker_id, None, worker_total_num)

        return iterator


# Your code
batch_list = []
batch_size = 40

# Create dataset
dataset = PartialHIVDataset(max_samples=batch_size)
iterator = iter(dataset)
full_batch_data = list(dataset)

df = pd.DataFrame(full_batch_data)

Initializing HIV dataset (will only load 40 samples)...


Processing...
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'O=C1O[Al]23(OC1=O)(OC(=O)C(=O)O2)OC(=O)C(=O)O3' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'Cc1ccc([B-2]2(c3ccc(C)cc3)=NCCO2)cc1' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'Oc1ccc(C2Oc3cc(O)cc4c3C(=[O+][AlH3-3]35([O+]=C6c7c(cc(O)cc7[OH+]3)OC(c3ccc(O)cc3O)C6O)([O+]=C3c6c(cc(O)cc6[OH+]5)OC(c5ccc(O)cc5O)C3O)[OH+]4)C2O)c(O)c1' since it resulted in zero atoms
  warnings.warn(f"Skipping molecule '{smiles}' since it "
/usr/local/lib/python3.12/dist-packages/torch_geometric/datasets/molecule_net.py:213: UserWarning: Skipping molecule 'CC1=C2

Dataset ready! Total size: 41120 molecules
But we'll only load 40 of them.



In [3]:
import codecs
import numpy as np
import torch.nn as nn
from transformers import AutoTokenizer

smiles_list = []
batch_size = 40

for _ in range(batch_size):
  mol_info = next(iterator)
  smiles_list.append(mol_info['smiles'])

tokenizer = AutoTokenizer.from_pretrained("seyonec/ChemBERTa-zinc-base-v1")

tokenized_lengths = []
for smiles in smiles_list:
    tokens = tokenizer(smiles, padding=False, truncation=False)
    tokenized_lengths.append(len(tokens['input_ids']))

max_length = max(tokenized_lengths)
print(f"Maximum sequence length in data: {max_length}")

encoded = tokenizer(
    smiles_list,
    padding='max_length',
    truncation=True,
    max_length=max_length,
    return_tensors='pt'
)

embedding = nn.Embedding(
    num_embeddings=len(tokenizer),
    embedding_dim=512,
    padding_idx=tokenizer.pad_token_id
)

embeddings = embedding(encoded['input_ids'])
print(f"Embeddings shape: {embeddings.shape}")  # (batch_size, max_length, 512)



/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/166 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/501 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/150 [00:00<?, ?B/s]

Maximum sequence length in data: 61
Embeddings shape: torch.Size([40, 61, 512])


In [4]:
from scipy.sparse.linalg import eigsh
from rdkit import Chem
def smiles_to_adjacency_matrix(smiles):
    """Convert SMILES to adjacency matrix."""
    mol = Chem.MolFromSmiles(smiles)
    if mol is None:
        return None

    n_atoms = mol.GetNumAtoms()
    adj_matrix = np.zeros((n_atoms, n_atoms))

    for bond in mol.GetBonds():
        i = bond.GetBeginAtomIdx()
        j = bond.GetEndAtomIdx()
        # Undirected graph
        adj_matrix[i, j] = 1
        adj_matrix[j, i] = 1

    return adj_matrix
#k is the number of eigenvectors
def compute_graph_positional_encoding(adj_matrix, k=30):
    """
    Compute graph positional encodings from eigenvectors of the
    symmetrically normalized graph Laplacian.
    """
    n = adj_matrix.shape[0]

    # Compute degree matrix
    degree = np.sum(adj_matrix, axis=1)
    degree[degree == 0] = 1

    # D^(-1/2)
    d_inv_sqrt = np.diag(1.0 / np.sqrt(degree))

    # Symmetrically normalized Laplacian
    identity = np.eye(n)
    normalized_adj = d_inv_sqrt @ adj_matrix @ d_inv_sqrt
    laplacian = identity - normalized_adj

    # Always use eigh for small matrices or when k >= n-1
    if n <= k + 1:  # Changed condition
        eigenvalues, eigenvectors = np.linalg.eigh(laplacian)
        if n < k:
            # Pad with zeros
            padded_eigenvectors = np.zeros((n, k))
            padded_eigenvectors[:, :n] = eigenvectors
            return padded_eigenvectors
        else:
            return eigenvectors[:, :k]
    else:
        # Use sparse solver only when k < n-1
        eigenvalues, eigenvectors = eigsh(laplacian, k=k, which='SM')
        return eigenvectors

In [5]:
# Process all SMILES and compute their graph PEs
graph_pes_list = []
max_nodes = 0

for smiles in smiles_list:
    adj_matrix = smiles_to_adjacency_matrix(smiles)
    if adj_matrix is not None:
        pe = compute_graph_positional_encoding(adj_matrix, k=30)
        graph_pes_list.append(pe)
        max_nodes = max(max_nodes, pe.shape[0])
    else:
        graph_pes_list.append(None)

print(f"Maximum number of atoms in dataset: {max_nodes}")


Maximum number of atoms in dataset: 42


In [6]:
import torch
import numpy as np
from rdkit import Chem
import re



def generate_zero_pe(n_vectors, dim=30):
    """Generate n_vectors zero vectors of dimension dim for non-atom characters."""
    return np.zeros((n_vectors, dim))

def parse_token_components(token):
    """
    Parse a token to identify atoms and characters.
    Returns: (atom_indices, n_characters)
    """
    # Common atom patterns in SMILES
    atom_pattern = r'(Cl|Br|Si|Mg|Ca|Fe|Al|Na|Li|[BCNOFPSKHIV])'

    # Find all atoms in the token
    atoms = re.findall(atom_pattern, token)

    # Count non-atom characters
    # Remove atoms from token to count remaining characters
    remaining = token
    for atom in atoms:
        remaining = remaining.replace(atom, '', 1)

    # Count actual characters (digits, +, -, =, #, etc.)
    n_characters = len([c for c in remaining])

    return atoms, n_characters

def align_graph_pe_to_tokens(smiles, graph_pe, tokenizer, max_length, random_seed=None):
    """
    Align graph PE to tokens, handling pure characters, atoms, and mixed tokens.
    """
    if random_seed is not None:
        np.random.seed(random_seed)

    # Determine embedding dimension
    if graph_pe is not None:
        embedding_dim = graph_pe.shape[1]
    else:
        embedding_dim = 30  # Default fallback

    mol = Chem.MolFromSmiles(smiles)
    if mol is None or graph_pe is None:
        return torch.zeros(max_length, embedding_dim)

    # Get tokens
    encoding = tokenizer(smiles, padding='max_length', max_length=max_length)
    tokens = tokenizer.convert_ids_to_tokens(encoding['input_ids'])
    token_pe = torch.zeros(len(tokens), embedding_dim)

    # Build atom mapping from SMILES
    atom_symbols = [atom.GetSymbol() for atom in mol.GetAtoms()]
    atom_count = {symbol: 0 for symbol in set(atom_symbols)}
    atom_to_idx = {}

    for idx, symbol in enumerate(atom_symbols):
        atom_to_idx[(symbol, atom_count[symbol])] = idx
        atom_count[symbol] += 1

    # Reset atom count for tracking
    current_atom_count = {symbol: 0 for symbol in atom_count}

    # Process each token
    for i, token in enumerate(tokens):
        if token in ['<s>', '</s>', '<pad>']:
            continue

        # Parse token components
        atoms_in_token, n_characters = parse_token_components(token)

        if len(atoms_in_token) == 0 and n_characters > 0:
            # Pure character token - use zero PE
            random_pes = generate_zero_pe(n_characters, embedding_dim)
            token_pe[i] = torch.tensor(random_pes.sum(axis=0))

        elif len(atoms_in_token) > 0 and n_characters == 0:
            # Pure atom token(s) - use graph PE
            atom_pes = []
            for atom_symbol in atoms_in_token:
                if atom_symbol in current_atom_count:
                    atom_key = (atom_symbol, current_atom_count[atom_symbol])
                    if atom_key in atom_to_idx:
                        atom_idx = atom_to_idx[atom_key]
                        atom_pes.append(graph_pe[atom_idx])
                        current_atom_count[atom_symbol] += 1

            if atom_pes:
                # sum the PEs of all atoms in this token
                token_pe[i] = torch.tensor(np.sum(atom_pes, axis=0))

        elif len(atoms_in_token) > 0 and n_characters > 0:
            # Mixed token - combine atom PE and character PE
            atom_pes = []
            for atom_symbol in atoms_in_token:
                if atom_symbol in current_atom_count:
                    atom_key = (atom_symbol, current_atom_count[atom_symbol])
                    if atom_key in atom_to_idx:
                        atom_idx = atom_to_idx[atom_key]
                        atom_pes.append(graph_pe[atom_idx])
                        current_atom_count[atom_symbol] += 1

            # Get character PEs (zeros as you intended)
            random_pes = generate_zero_pe(n_characters, embedding_dim)

            # Combine: sum of atoms + sum of characters
            combined_pe = np.zeros(embedding_dim)
            if atom_pes:
                combined_pe += np.sum(atom_pes, axis=0)
            combined_pe += random_pes.sum(axis=0)

            token_pe[i] = torch.tensor(combined_pe)

    # Now pad or truncate to max_length
    if len(tokens) < max_length:
        # Pad with zeros
        padded_token_pe = torch.zeros(max_length, embedding_dim)
        padded_token_pe[:len(tokens)] = token_pe
        return padded_token_pe
    else:
        # Truncate
        return token_pe[:max_length]



In [7]:
token_pes_list = []

for smiles, graph_pe in zip(smiles_list, graph_pes_list):
    # Align graph PE to tokens using the provided function
    token_pe = align_graph_pe_to_tokens(
        smiles=smiles,
        graph_pe=graph_pe,
        tokenizer=tokenizer,
        max_length=max_length,
        random_seed=42
    )
    token_pes_list.append(token_pe)

# Stack into batch tensor
token_pes_batch = torch.stack(token_pes_list)

In [8]:
class EmbeddingWithGraphPE(nn.Module):
    def __init__(self, embed_dim=512, pe_dim=30):
        super().__init__()
        # One-layer projection with GeLU from graph PE to embedding dimension
        # Using standard Laplacian, so pe_dim is just k
        self.pe_projection = nn.Sequential(
            nn.Linear(pe_dim, embed_dim),
            nn.GELU()
        )

    def forward(self, embeddings, token_pes):
        # token_pes shape: [batch_size, seq_len, pe_dim]
        # embeddings shape: [batch_size, seq_len, embed_dim]

        # Project token PEs to embedding dimension with GeLU
        projected_pes = self.pe_projection(token_pes)

        # Add to token embeddings
        enhanced_embeddings = embeddings + projected_pes

        return enhanced_embeddings

# Usage
model = EmbeddingWithGraphPE(embed_dim=512, pe_dim=30)
# graph_pes_batch shape: [40, 61, 30] (30 eigenvectors from standard Laplacian)
# embeddings shape: [40, 61, 512]
enhanced_embeddings = model(embeddings, token_pes_batch)
# Output shape: [40, 61, 512]

In [9]:
enhanced_embeddings.shape

torch.Size([40, 61, 512])

In [10]:
import json
from data_import import check_class_balance, stratified_train_val_test_split, convert_to_litgpt_format, save_to_jsonl

# Create dataset
batch_size = 41120
dataset = PartialHIVDataset(max_samples=batch_size)

# Convert to DataFrame
full_batch_data = list(dataset)
df = pd.DataFrame(full_batch_data)

# Check balance in the original DataFrame
original_balance = check_class_balance(df, "Original Dataset")

# Convert to LitGPT format for different task types
classification_data = convert_to_litgpt_format(df, task_type="classification")

# Check balance in the converted data
converted_balance = check_class_balance(classification_data, "Converted Dataset")

# Create stratified train/validation/test split (60/20/20)
train_data, val_data, test_data = stratified_train_val_test_split(
    classification_data,
    train_ratio=0.6,
    val_ratio=0.2,
    test_ratio=0.2
)

# Check balance in all sets
train_balance = check_class_balance(train_data, "Training Set")
val_balance = check_class_balance(val_data, "Validation Set")
test_balance = check_class_balance(test_data, "Test Set")

# Save to JSONL files
save_to_jsonl(train_data, "hiv_train.jsonl")
save_to_jsonl(val_data, "hiv_val.jsonl")
save_to_jsonl(test_data, "hiv_test.jsonl")

print(f"\nDataset sizes:")
print(f"Train: {len(train_data)}")
print(f"Val: {len(val_data)}")
print(f"Test: {len(test_data)}")

Initializing HIV dataset (will only load 41120 samples)...
Dataset ready! Total size: 41120 molecules
But we'll only load 41120 of them.


Original Dataset Class Balance:
Total samples: 41120
Class 0.0: 39677 samples (96.49%)
Class 1.0: 1443 samples (3.51%)

Converted Dataset Class Balance:
Total samples: 41120
Class 0: 39677 samples (96.49%)
Class 1: 1443 samples (3.51%)

Training Set Class Balance:
Total samples: 24671
Class 0: 23806 samples (96.49%)
Class 1: 865 samples (3.51%)

Validation Set Class Balance:
Total samples: 8224
Class 0: 7935 samples (96.49%)
Class 1: 289 samples (3.51%)

Test Set Class Balance:
Total samples: 8225
Class 0: 7936 samples (96.49%)
Class 1: 289 samples (3.51%)
Saved 24671 entries to hiv_train.jsonl
Saved 8224 entries to hiv_val.jsonl
Saved 8225 entries to hiv_test.jsonl

Dataset sizes:
Train: 24671
Val: 8224
Test: 8225


In [11]:
import json
import torch
from utils import calculate_max_lengths
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer

def load_jsonl(filepath):
    """Load data from JSONL file"""
    data = []
    with open(filepath, 'r') as f:
        for line in f:
            data.append(json.loads(line))
    return data

class HIVDataset(Dataset):
    def __init__(self, jsonl_filepath, tokenizer, max_full_length):
        self.data = load_jsonl(jsonl_filepath)
        self.tokenizer = tokenizer
        self.max_full_length = max_full_length

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        item = self.data[idx]

        # Extract SMILES
        smiles = item['input'].replace("SMILES: ", "")

        # Create full prompt
        full_prompt = f"{item['instruction']}\n{item['input']}\nAnswer:"
        # Compute graph PE with actual SMILES length
        smiles_tokens = self.tokenizer(smiles, add_special_tokens=False)
        actual_smiles_length = len(smiles_tokens['input_ids'])
        # Get prompt length for label masking
        prompt_encoding = self.tokenizer(full_prompt, add_special_tokens=True, return_tensors='pt')
        prompt_length = prompt_encoding['input_ids'].shape[1]

        # Full sequence for training
        target_text = f"{full_prompt} {item['output']}"
        full_encoding = self.tokenizer(
            target_text,
            padding='max_length',
            max_length=self.max_full_length,
            truncation=True,
            return_tensors='pt'
        )

        # Compute graph PE for SMILES only
        adj_matrix = smiles_to_adjacency_matrix(smiles)
        if adj_matrix is not None:
            graph_pe = compute_graph_positional_encoding(adj_matrix, k=30)

            smiles_token_pe = align_graph_pe_to_tokens(
                smiles, graph_pe, self.tokenizer, actual_smiles_length, random_seed=42
            )
        else:
            smiles_token_pe = torch.zeros(actual_smiles_length, 30)

        # Create full-length PE tensor
        full_token_pe = torch.zeros(self.max_full_length, 30)

        # Find where SMILES tokens are and place PEs there
        instruction_text = f"{item['instruction']}\nSMILES: "
        instruction_tokens = self.tokenizer(instruction_text, add_special_tokens=True)
        smiles_start_idx = len(instruction_tokens['input_ids'])

        # Copy SMILES PEs to correct position
        pe_end_idx = min(smiles_start_idx + actual_smiles_length, self.max_full_length)
        pe_length = min(smiles_token_pe.shape[0], pe_end_idx - smiles_start_idx)
        full_token_pe[smiles_start_idx:smiles_start_idx + pe_length] = smiles_token_pe[:pe_length]
        # Debug: Check final PE tensor

        # Create labels with prompt masked
        labels = full_encoding['input_ids'][0].clone()
        labels[:prompt_length] = -100  # Mask prompt tokens
        # Also mask padding tokens
        attention_mask = full_encoding['attention_mask'][0]
        labels[attention_mask == 0] = -100
        return {
            'input_ids': full_encoding['input_ids'][0],
            'graph_pes': full_token_pe,
            'labels': labels,
        }

# First calculate max lengths
max_full_length = calculate_max_lengths('hiv_train.jsonl')

def create_dataloaders(train_path='hiv_train.jsonl',
                      val_path='hiv_val.jsonl',
                      test_path='hiv_test.jsonl',  # Add test path
                      batch_size=8,
                      tokenizer=None,
                      max_full_length=max_full_length):
    if tokenizer is None:
        tokenizer = AutoTokenizer.from_pretrained("EleutherAI/pythia-160m")
        if tokenizer.pad_token is None:
            tokenizer.pad_token = tokenizer.eos_token

    train_dataset = HIVDataset(train_path, tokenizer, max_full_length)
    val_dataset = HIVDataset(val_path, tokenizer, max_full_length)
    test_dataset = HIVDataset(test_path, tokenizer, max_full_length)  # Add test dataset

    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=batch_size, shuffle=False)
    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)  # Add test loader

    return train_loader, val_loader, test_loader, tokenizer  # Return test_loader too

# Updated usage
train_loader, val_loader, test_loader, tokenizer = create_dataloaders(batch_size=8)

# Test one batch from each loader
print("Train batch:")
sample_batch = next(iter(train_loader))
print(f"Input IDs shape: {sample_batch['input_ids'].shape}")
print(f"Graph PEs shape: {sample_batch['graph_pes'].shape}")
print(f"Labels shape: {sample_batch['labels'].shape}")

tokenizer_config.json: 0.00B [00:00, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/396 [00:00<?, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/99.0 [00:00<?, ?B/s]

Train batch:
Input IDs shape: torch.Size([8, 498])
Graph PEs shape: torch.Size([8, 498, 30])
Labels shape: torch.Size([8, 498])


In [12]:
# Look at first example in batch
i = 0
print(f"Input text (decoded): {tokenizer.decode(sample_batch['input_ids'][i])}")
print(f"\nNon-zero PE positions: {torch.nonzero(sample_batch['graph_pes'][i].sum(dim=1)).squeeze().tolist()}")
print(f"\nLabel tokens that aren't -100: {sample_batch['labels'][i][sample_batch['labels'][i] != -100].tolist()}")

Input text (decoded): Classify the following molecule based on its HIV activity. Respond with '1' if the molecule shows HIV activity, or '0' if it does not.
SMILES: O=C1C2C(C(=O)N1c1ccccc1)C1([N+](=O)[O-])C=C3c4ccccc4N(C(=O)c4ccccc4)C32C2C(=O)N(c3ccccc3)C(=O)C21
Answer: 0<|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|><|endoftext|>



---

## Implementation, Results, Comments and Challenges

### 1. Design Decisions and Simplifications

#### Handling Non-Atom Characters
- Set positional encodings (PEs) for non-atom characters to zero
- Preserves structural information from atoms while avoiding noise from SMILES syntax

#### Addressing Class Imbalance
Dataset: 96.5% negative vs 3.5% positive HIV activity

Experimented with three loss functions:
| Loss Function | Result |
|---------------|---------|
| Focal loss | Unstable |
| Dice loss | Produced inf values |
| **Weighted cross-entropy** | **Most stable (selected)** |

#### Training Configuration
- Learning rate warm-up following paper methodology
- Stratified train/val/test split preserving original distribution
- Early stopping at 250 batches for rapid experimentation

### 2. Alignment Function Architecture

Token-to-PE mapping handles three cases:

```
Pure character tokens    →  Zero PE
Pure atom tokens        →  Graph PE from atom
Mixed tokens           →  Sum(atom PEs) + Sum(zero PEs)
```

**Future consideration:** Functional group-level patterns could be enforced, though this risks over-engineering features the network should learn independently.

### 3. Results Analysis

#### Performance Summary

| Model | Class 0 Acc | Class 1 Acc | Test Loss | F1 Score |
|-------|-------------|-------------|-----------|----------|
| **With PE** | 99.99% | 100% | 0.0066 | 0.9983 |
| **Without PE** | 100% | 100% | 0.0063 | 1.0000 |

#### Key Findings

**1. Convergence Speed**
- Both models reach >99% accuracy within 200-250 batches
- Loss drops to <0.01 in 200 iterations
- Suggests task complexity may be insufficient for architecture. This was coroborated when I increased the L2 regularization weight decay to 0.5 to see if there was a degradation in performance. I hypothesized that a higher weight decay should prevent it from learning a complex pattern but the performance remained near perfect. This seems to indicate the task is too easy rather than a case of overfitting. I also checked the activations and they didnt seem to be abnormal:

Layer 0: mean=-0.000, std=0.003, % dead neurons=0.000
Layer 1: mean=0.014, std=1.774, % dead neurons=0.000
Layer 2: mean=1.604, std=3.106, % dead neurons=0.000
etc...

**2. PE Impact**
- No measurable improvement from positional encodings

**3. Possible Explanations**
- HIV activity may depend on simple functional group presence/absence
- SMILES sequential representation already encodes connectivity

#### Validity Concerns

Despite suspicious perfect accuracy on imbalanced data:
- Strong performance on unseen test set contradicts overfitting
- Verified no data leakage between train/val/test splits

# Technical Challenges from HIV Molecular Property Prediction Task

## 1. Graph PE to Token Alignment
**Challenge**: Tokenizer splits molecules unpredictably (e.g., 'Cl' → ['C', 'l'] or ['Cl'])
**Solution**: Built regex parser to identify atoms within tokens:
```python
atom_pattern = r'(Cl|Br|Si|Mg|Ca|Fe|Al|Na|Li|[BCNOFPSKHIV])'
# Three cases: pure chars → zero PE, pure atoms → graph PE, mixed → sum both
```

## 2. Extreme Class Imbalance (96.5% vs 3.5%)
**Challenge**: Models predicted all negatives for easy 96.5% accuracy
**Solution**:
- Tested focal loss (unstable), dice loss (inf values), settled on weighted CE [0.035, 0.965]
- Stratified sampling + per-class metrics tracking

## 3. Suspicious Perfect Performance
**Challenge**: >99% accuracy in 200 batches seemed unrealistic
**Solution**: Diagnostic testing revealed genuine simplicity:
- L2=0.5 → no degradation (not overfitting)
- No dead neurons in activation analysis
- Verified no data leakage via stratified splits
- Consistent performance on held-out test set

## 4. GPU Memory Management
**Challenge**: Sequential model training caused OOM errors
**Solution**:
```python
del model_with_pe; torch.cuda.empty_cache(); gc.collect()
```
Save checkpoints → reload for evaluation

## 5. Silent PE Alignment Failures
**Challenge**: Some molecules produced all-zero encodings without errors
**Solution**: Added debugging + fallback behavior:
```python
print(f"Non-zero PE positions: {torch.nonzero(graph_pes.sum(dim=1))}")
```

## 6. Eigendecomposition Instability
**Challenge**: Sparse eigensolver failed for tiny molecules (1-2 atoms)
**Solution**: Conditional logic based on matrix size:
```python
if n <= k + 1:  # Dense solver for small molecules
    eigenvectors = np.linalg.eigh(laplacian)
    if n < k: pad_with_zeros()

<div align="center">
<img src="https://drive.google.com/uc?id=1lW4OBGNmaqacNgSU03aK7ym0n9jxIicG" alt="Results visualization" width="900"/>
</div>

In [26]:
import torch
import torch.nn.functional as F
import torch.nn as nn
from transformers import AutoModelForCausalLM, AutoTokenizer,get_linear_schedule_with_warmup
from peft import LoraConfig, get_peft_model, TaskType
from typing import Optional

import matplotlib.pyplot as plt

import torch
import torch.nn.functional as F
import torch.nn as nn
from transformers import AutoModelForCausalLM, AutoTokenizer, get_linear_schedule_with_warmup
from peft import LoraConfig, get_peft_model, TaskType
from typing import Optional

class HIVPEModule(nn.Module):
    def __init__(self, pe_dim=30, embed_dim=768):
        super().__init__()
        self.pe_projection = nn.Linear(pe_dim, embed_dim)
        self.activation = nn.GELU()
        self.dropout = nn.Dropout(0.2)
        nn.init.xavier_uniform_(self.pe_projection.weight, gain=0.3)
        nn.init.zeros_(self.pe_projection.bias)
        self.scale = 1

    def forward(self, embeddings, graph_pes):
        if graph_pes is not None:
            graph_pes = graph_pes.to(embeddings.dtype)
            projected_pes = self.pe_projection(graph_pes)
            projected_pes = self.activation(projected_pes)
            projected_pes = self.dropout(projected_pes)
            return embeddings + self.scale * projected_pes
        return embeddings

class PythiaWithPE(nn.Module):
    def __init__(self, base_model, pe_dim=30,use_pe=True):
        super().__init__()
        self.base_model = base_model
        self.embed_dim = base_model.gpt_neox.embed_in.embedding_dim
        self.use_pe = use_pe
        if self.use_pe:
          self.pe_module = HIVPEModule(pe_dim=pe_dim, embed_dim=self.embed_dim)

    def forward(self, input_ids=None, graph_pes=None, attention_mask=None, labels=None, **kwargs):
        inputs_embeds = self.base_model.gpt_neox.embed_in(input_ids)

        if self.use_pe and graph_pes is not None:
            inputs_embeds = self.pe_module(inputs_embeds, graph_pes)

        kwargs.pop('input_ids', None)

        outputs = self.base_model(
            inputs_embeds=inputs_embeds,
            attention_mask=attention_mask,
            labels=labels,
            **kwargs
        )

        return outputs

    def generate(self, input_ids, graph_pes=None, **kwargs):
        inputs_embeds = self.base_model.gpt_neox.embed_in(input_ids)
        if self.use_pe and graph_pes is not None:
            inputs_embeds = self.pe_module(inputs_embeds, graph_pes)

        return self.base_model.generate(
            inputs_embeds=inputs_embeds,
            **kwargs
        )

def setup_model_with_pe_and_lora(model_name="EleutherAI/pythia-160m", pe_dim=30,use_pe=True):
    print(f"Loading {model_name}...")
    print(f"Using positional encodings: {use_pe}")
    base_model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype=torch.float32,
        device_map="auto"
    )

    tokenizer = AutoTokenizer.from_pretrained(model_name)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    model = PythiaWithPE(base_model, pe_dim=pe_dim, use_pe=use_pe)

    lora_config = LoraConfig(
        task_type=TaskType.CAUSAL_LM,
        r=4,
        lora_alpha=8,
        lora_dropout=0.3,
        target_modules=["query_key_value", "dense_h_to_4h", "dense_4h_to_h"],
        init_lora_weights="gaussian"
    )

    model.base_model = get_peft_model(model.base_model, lora_config)

    total_params = sum(p.numel() for p in model.parameters())
    trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print(f"Total parameters: {total_params:,}")
    print(f"Trainable parameters: {trainable_params:,}")
    print(f"Trainable %: {100 * trainable_params / total_params:.2f}%")

    return model, tokenizer
def quick_evaluate(model, data_loader, split_name="Quick Val", max_batches=50):
      model.eval()
      total_loss = 0
      class_0_correct = 0
      class_1_correct = 0
      class_0_total = 0
      class_1_total = 0
      num_batches = 0
      # Define tokens here
      token_0 = 470  # " 0"
      token_1 = 337  # " 1"
      device = next(model.parameters()).device

      with torch.no_grad():
          for batch_idx, batch in enumerate(data_loader):
              if batch_idx >= max_batches:  # Only evaluate first N batches
                  break

              input_ids = batch['input_ids'].to(device)
              graph_pes = batch['graph_pes'].to(device)
              labels = batch['labels'].to(device)

              outputs = model(input_ids=input_ids, graph_pes=graph_pes)
              logits = outputs.logits

              mask = ((labels == token_0) | (labels == token_1)) & (labels != -100)

              if mask.any():
                  logits_flat = logits.view(-1, logits.size(-1))
                  labels_flat = labels.view(-1)
                  mask_flat = mask.view(-1)

                  relevant_logits = logits_flat[mask_flat]
                  relevant_labels = labels_flat[mask_flat]

                  loss = weighted_cross_entropy(relevant_logits, relevant_labels)
                  total_loss += loss.item()
                  num_batches += 1

                  preds = relevant_logits.argmax(dim=-1)

                  class_0_mask = relevant_labels == token_0
                  class_1_mask = relevant_labels == token_1

                  if class_0_mask.any():
                      class_0_correct += (preds[class_0_mask] == relevant_labels[class_0_mask]).sum().item()
                      class_0_total += class_0_mask.sum().item()

                  if class_1_mask.any():
                      class_1_correct += (preds[class_1_mask] == relevant_labels[class_1_mask]).sum().item()
                      class_1_total += class_1_mask.sum().item()

      model.train()

      avg_loss = total_loss / num_batches if num_batches > 0 else float('inf')
      class_0_acc = class_0_correct / class_0_total if class_0_total > 0 else 0
      class_1_acc = class_1_correct / class_1_total if class_1_total > 0 else 0

      print(f"  {split_name} (first {max_batches} batches): Loss: {avg_loss:.4f}, "
            f"Class 0: {class_0_acc:.2%}, Class 1: {class_1_acc:.2%}")

      return avg_loss, class_0_acc, class_1_acc


def weighted_cross_entropy(logits, labels):
        token_0 = 470  # " 0"
        token_1 = 337  # " 1"
        ce_loss = F.cross_entropy(logits, labels, reduction='none')
        sample_weights = torch.where(
            labels == token_1,
            torch.tensor(0.965, device=logits.device),
            torch.tensor(0.035, device=logits.device)
        )
        return (ce_loss * sample_weights).mean()

def train_model(
    model,
    train_loader,
    val_loader=None,
    epochs=3,
    learning_rate=1e-5,
    device="cuda",
    tokenizer=None, eval_every=100, max_batches=250
):
    model = model.to(device)
    gradient_accumulation_steps = 2

    optimizer = torch.optim.AdamW(
        model.parameters(),
        lr=learning_rate,
        weight_decay=0.5,
        eps=1e-6
    )

    num_training_steps = (len(train_loader) * epochs) // gradient_accumulation_steps
    scheduler = get_linear_schedule_with_warmup(
        optimizer,
        num_warmup_steps=100,
        num_training_steps=num_training_steps
    )

    token_0 = 470  # " 0"
    token_1 = 337  # " 1"

    scaler = torch.amp.GradScaler('cuda') if device == "cuda" else None

    ###training###
    for epoch in range(epochs):
        model.train()
        total_loss = 0
        valid_batches = 0
        train_class_0_correct = 0
        train_class_1_correct = 0
        train_class_0_total = 0
        train_class_1_total = 0
        for batch_idx, batch in enumerate(train_loader):
            if batch_idx >= max_batches:
                  print(f"\nStopping training at batch {batch_idx} as requested")
                  break
            input_ids = batch['input_ids'].to(device)
            graph_pes = batch['graph_pes'].to(device)
            labels = batch['labels'].to(device)

            if scaler is not None:
                with torch.amp.autocast('cuda'):
                    outputs = model(input_ids=input_ids, graph_pes=graph_pes)
                    logits = outputs.logits

                    if torch.isnan(logits).any():
                        optimizer.zero_grad()
                        continue

                    mask = ((labels == token_0) | (labels == token_1)) & (labels != -100)

                    if mask.any():
                        logits_flat = logits.view(-1, logits.size(-1))
                        labels_flat = labels.view(-1)
                        mask_flat = mask.view(-1)

                        relevant_logits = logits_flat[mask_flat]
                        relevant_labels = labels_flat[mask_flat]

                        loss = weighted_cross_entropy(relevant_logits, relevant_labels)
                        # logit_reg = 0.001 * (relevant_logits ** 2).mean()
                        # loss = loss + logit_reg
                        loss = loss / gradient_accumulation_steps
                        with torch.no_grad():
                          preds = relevant_logits.argmax(dim=-1)

                          # Track per-class accuracy
                          class_0_mask = relevant_labels == token_0
                          class_1_mask = relevant_labels == token_1

                          if class_0_mask.any():
                              train_class_0_correct += (preds[class_0_mask] == relevant_labels[class_0_mask]).sum().item()
                              train_class_0_total += class_0_mask.sum().item()

                          if class_1_mask.any():
                              train_class_1_correct += (preds[class_1_mask] == relevant_labels[class_1_mask]).sum().item()
                              train_class_1_total += class_1_mask.sum().item()

                    else:
                        continue

                scaler.scale(loss).backward()

                if (batch_idx + 1) % gradient_accumulation_steps == 0:
                    scaler.unscale_(optimizer)
                    torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)
                    scaler.step(optimizer)
                    scaler.update()
                    optimizer.zero_grad()
                    scheduler.step()

                if mask.any() and not torch.isnan(loss):
                    total_loss += loss.item() * gradient_accumulation_steps
                    valid_batches += 1

                if batch_idx % 50 == 0:

                    class_0_acc = train_class_0_correct / train_class_0_total if train_class_0_total > 0 else 0
                    class_1_acc = train_class_1_correct / train_class_1_total if train_class_1_total > 0 else 0

                    print(f"Epoch {epoch+1}/{epochs}, Batch {batch_idx}/{len(train_loader)}, "
                          f"Loss: {loss.item() * gradient_accumulation_steps:.4f}")
                    print(f"  Train Class 0 acc: {class_0_acc:.2%} ({train_class_0_correct}/{train_class_0_total})")
                    if train_class_1_total > 0:
                        print(f"  Train Class 1 acc: {class_1_acc:.2%} ({train_class_1_correct}/{train_class_1_total})")
                    else:
                        print(f"  Train Class 1 acc: No samples in last 50 batches")

                    # Reset counters
                    train_class_0_correct = 0
                    train_class_1_correct = 0
                    train_class_0_total = 0
                    train_class_1_total = 0
                if batch_idx % eval_every == 0 and batch_idx > 0 and val_loader is not None:
                  print(f"\nMid-epoch evaluation at batch {batch_idx}:")
                  quick_evaluate(model, val_loader, "Validation")
                  model.train()  # Switch back to training mode
        print(f"\nEpoch {epoch+1} Summary:")
        print(f"Average Loss: {total_loss/valid_batches:.4f}")

        if val_loader is not None:
            quick_evaluate(model, val_loader, "Validation")

    return model

# Updated usage code
if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"

    print("Training WITH positional encodings:")
    model_with_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=True)
    model_with_pe = model_with_pe.to(device)

    print("\nTraining WITHOUT positional encodings:")
    model_without_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=False)
    model_without_pe = model_without_pe.to(device)

    print(f"\nModels setup complete and moved to {device}!")
    print("You can now train both models and compare their performance.")

Training WITH positional encodings:
Loading EleutherAI/pythia-160m...
Using positional encodings: True
Total parameters: 162,862,848
Trainable parameters: 539,904
Trainable %: 0.33%

Training WITHOUT positional encodings:
Loading EleutherAI/pythia-160m...
Using positional encodings: False
Total parameters: 162,839,040
Trainable parameters: 516,096
Trainable %: 0.32%

Models setup complete and moved to cuda!
You can now train both models and compare their performance.


In [27]:
import gc
import torch

# Clear GPU memory before starting
torch.cuda.empty_cache()
gc.collect()

# Setup device
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"Using device: {device}")

# Training configuration
BATCH_SIZE = 10
EPOCHS = 1
LEARNING_RATE = 1e-4
MAX_BATCHES = 250  # Set to None for full training

# Create dataloaders (same for both models)
print("\nCreating dataloaders...")
train_loader, val_loader, test_loader, tokenizer = create_dataloaders(
    batch_size=BATCH_SIZE,
    tokenizer=None  # Will use tokenizer from model setup
)

# Check class distribution
print("\nChecking class 1 presence in first 10 batches:")
class_1_count = 0
for i, batch in enumerate(train_loader):
    if i >= 10:
        break
    labels = batch['labels']
    if (labels == 337).any():
        class_1_count += 1
print(f"Found class 1 in {class_1_count}/10 batches")

# Store results for comparison
results = {}

# Train model WITH positional encodings
print("\n" + "="*50)
print("TRAINING MODEL WITH POSITIONAL ENCODINGS")
print("="*50)

model_with_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=True)
model_with_pe = model_with_pe.to(device)

trained_model_with_pe = train_model(
    model_with_pe,
    train_loader,
    val_loader,
    epochs=EPOCHS,
    learning_rate=LEARNING_RATE,
    device=device,
    tokenizer=tokenizer,
    max_batches=MAX_BATCHES
)

# Evaluate model with PE
print("\nFinal evaluation WITH PE:")
loss_with_pe, class0_with_pe, class1_with_pe = quick_evaluate(
    trained_model_with_pe,
    test_loader,
    "Test Set WITH PE",
    max_batches=100
)

results['with_pe'] = {
    'loss': loss_with_pe,
    'class_0_acc': class0_with_pe,
    'class_1_acc': class1_with_pe
}
torch.save(trained_model_with_pe.state_dict(), 'model_with_pe.pth')
# Clear GPU memory between models
del model_with_pe
del trained_model_with_pe
torch.cuda.empty_cache()
gc.collect()

# Train model WITHOUT positional encodings
print("\n" + "="*50)
print("TRAINING MODEL WITHOUT POSITIONAL ENCODINGS")
print("="*50)

model_without_pe, tokenizer = setup_model_with_pe_and_lora(use_pe=False)
model_without_pe = model_without_pe.to(device)

trained_model_without_pe = train_model(
    model_without_pe,
    train_loader,
    val_loader,
    epochs=EPOCHS,
    learning_rate=LEARNING_RATE,
    device=device,
    tokenizer=tokenizer,
    max_batches=MAX_BATCHES
)

# Evaluate model without PE
print("\nFinal evaluation WITHOUT PE:")
loss_without_pe, class0_without_pe, class1_without_pe = quick_evaluate(
    trained_model_without_pe,
    test_loader,
    "Test Set WITHOUT PE",
    max_batches=100
)

results['without_pe'] = {
    'loss': loss_without_pe,
    'class_0_acc': class0_without_pe,
    'class_1_acc': class1_without_pe
}
torch.save(trained_model_without_pe.state_dict(), 'model_without_pe.pth')
# Compare results
print("\n" + "="*50)
print("COMPARISON OF RESULTS")
print("="*50)
print(f"\nModel WITH Positional Encodings:")
print(f"  Test Loss: {results['with_pe']['loss']:.4f}")
print(f"  Class 0 Accuracy: {results['with_pe']['class_0_acc']:.2%}")
print(f"  Class 1 Accuracy: {results['with_pe']['class_1_acc']:.2%}")

print(f"\nModel WITHOUT Positional Encodings:")
print(f"  Test Loss: {results['without_pe']['loss']:.4f}")
print(f"  Class 0 Accuracy: {results['without_pe']['class_0_acc']:.2%}")
print(f"  Class 1 Accuracy: {results['without_pe']['class_1_acc']:.2%}")

print(f"\nDifferences (WITH PE - WITHOUT PE):")
print(f"  Loss Difference: {results['with_pe']['loss'] - results['without_pe']['loss']:.4f}")
print(f"  Class 0 Acc Difference: {results['with_pe']['class_0_acc'] - results['without_pe']['class_0_acc']:.2%}")
print(f"  Class 1 Acc Difference: {results['with_pe']['class_1_acc'] - results['without_pe']['class_1_acc']:.2%}")


Using device: cuda

Creating dataloaders...

Checking class 1 presence in first 10 batches:
Found class 1 in 4/10 batches

TRAINING MODEL WITH POSITIONAL ENCODINGS
Loading EleutherAI/pythia-160m...
Using positional encodings: True
Total parameters: 162,862,848
Trainable parameters: 539,904
Trainable %: 0.33%
Epoch 1/1, Batch 0/2468, Loss: 2.1118
  Train Class 0 acc: 0.00% (0/7)
  Train Class 1 acc: 0.00% (0/3)
Epoch 1/1, Batch 50/2468, Loss: 0.0054
  Train Class 0 acc: 25.00% (120/480)
  Train Class 1 acc: 5.00% (1/20)
Epoch 1/1, Batch 100/2468, Loss: 0.0000
  Train Class 0 acc: 100.00% (478/478)
  Train Class 1 acc: 72.73% (16/22)

Mid-epoch evaluation at batch 100:
  Validation (first 50 batches): Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%
Epoch 1/1, Batch 150/2468, Loss: 0.0000
  Train Class 0 acc: 100.00% (477/477)
  Train Class 1 acc: 100.00% (23/23)


KeyboardInterrupt: 

In [19]:
#We analyze the activations to check for dead neurons and abnormal values, we notice high
def analyze_activations(model, data_loader):
    activations = []

    def hook_fn(module, input, output):
        activations.append(output.detach())

    # Register hooks on key layers
    hooks = []
    for name, module in model.named_modules():
        if isinstance(module, nn.Linear):
            hooks.append(module.register_forward_hook(hook_fn))

    # Get device from model
    device = next(model.parameters()).device

    # Run one batch
    batch = next(iter(data_loader))
    with torch.no_grad():
        # Move data to same device as model
        input_ids = batch['input_ids'].to(device)
        graph_pes = batch['graph_pes'].to(device)
        model(input_ids, graph_pes)

    # Analyze
    for i, act in enumerate(activations):
        print(f"Layer {i}: mean={act.mean():.3f}, std={act.std():.3f}, "
              f"% dead neurons={(act == 0).float().mean():.3f}")

    # Clean up hooks
    for hook in hooks:
        hook.remove()

# Now run it
analyze_activations(model_with_pe, test_loader)



Layer 0: mean=-0.000, std=0.003, % dead neurons=0.000
Layer 1: mean=0.014, std=1.774, % dead neurons=0.000
Layer 2: mean=1.604, std=3.106, % dead neurons=0.000
Layer 3: mean=-0.000, std=0.002, % dead neurons=0.000
Layer 4: mean=-0.002, std=0.251, % dead neurons=0.000
Layer 5: mean=-0.405, std=0.620, % dead neurons=0.000
Layer 6: mean=1.684, std=4.519, % dead neurons=0.000
Layer 7: mean=-0.000, std=0.004, % dead neurons=0.000
Layer 8: mean=-0.008, std=0.565, % dead neurons=0.000
Layer 9: mean=-0.871, std=5.044, % dead neurons=0.000
Layer 10: mean=0.000, std=0.004, % dead neurons=0.000
Layer 11: mean=-0.038, std=1.859, % dead neurons=0.000
Layer 12: mean=2.897, std=5.452, % dead neurons=0.000
Layer 13: mean=-0.000, std=0.005, % dead neurons=0.000
Layer 14: mean=0.001, std=0.172, % dead neurons=0.000
Layer 15: mean=-0.414, std=1.073, % dead neurons=0.000
Layer 16: mean=1.123, std=4.483, % dead neurons=0.000
Layer 17: mean=0.000, std=0.004, % dead neurons=0.000
Layer 18: mean=0.004, std=1.

In [30]:
def evaluate_with_progress(model, data_loader, split_name="Test Set", print_every=50):
    """
    Full evaluation with progress tracking
    """
    model.eval()
    total_loss = 0
    class_0_correct = 0
    class_1_correct = 0
    class_0_total = 0
    class_1_total = 0
    num_batches = 0
    total_batches = len(data_loader)

    # Define tokens
    token_0 = 470  # " 0"
    token_1 = 337  # " 1"
    device = next(model.parameters()).device

    # Define weighted_cross_entropy
    def weighted_cross_entropy(logits, labels):
        ce_loss = F.cross_entropy(logits, labels, reduction='none')
        sample_weights = torch.where(
            labels == token_1,
            torch.tensor(0.965, device=logits.device),
            torch.tensor(0.035, device=logits.device)
        )
        return (ce_loss * sample_weights).mean()

    print(f"\nStarting {split_name} evaluation on {total_batches} batches...")

    with torch.no_grad():
        for batch_idx, batch in enumerate(data_loader):
            input_ids = batch['input_ids'].to(device)
            graph_pes = batch['graph_pes'].to(device)
            labels = batch['labels'].to(device)

            outputs = model(input_ids=input_ids, graph_pes=graph_pes)
            logits = outputs.logits

            mask = ((labels == token_0) | (labels == token_1)) & (labels != -100)

            if mask.any():
                logits_flat = logits.view(-1, logits.size(-1))
                labels_flat = labels.view(-1)
                mask_flat = mask.view(-1)

                relevant_logits = logits_flat[mask_flat]
                relevant_labels = labels_flat[mask_flat]

                loss = weighted_cross_entropy(relevant_logits, relevant_labels)
                total_loss += loss.item()
                num_batches += 1

                preds = relevant_logits.argmax(dim=-1)

                class_0_mask = relevant_labels == token_0
                class_1_mask = relevant_labels == token_1

                if class_0_mask.any():
                    class_0_correct += (preds[class_0_mask] == relevant_labels[class_0_mask]).sum().item()
                    class_0_total += class_0_mask.sum().item()

                if class_1_mask.any():
                    class_1_correct += (preds[class_1_mask] == relevant_labels[class_1_mask]).sum().item()
                    class_1_total += class_1_mask.sum().item()

            # Print progress
            if (batch_idx + 1) % print_every == 0:
                progress = (batch_idx + 1) / total_batches * 100
                print(f"  Progress: {batch_idx + 1}/{total_batches} batches ({progress:.1f}%)")
                if num_batches > 0:
                    current_loss = total_loss / num_batches
                    current_acc_0 = class_0_correct / class_0_total if class_0_total > 0 else 0
                    current_acc_1 = class_1_correct / class_1_total if class_1_total > 0 else 0
                    print(f"  Current - Loss: {current_loss:.4f}, Class 0: {current_acc_0:.2%}, Class 1: {current_acc_1:.2%}")

    model.train()

    # Final results
    avg_loss = total_loss / num_batches if num_batches > 0 else float('inf')
    class_0_acc = class_0_correct / class_0_total if class_0_total > 0 else 0
    class_1_acc = class_1_correct / class_1_total if class_1_total > 0 else 0

    print(f"\n{split_name} Final Results:")
    print(f"  Loss: {avg_loss:.4f}")
    print(f"  Class 0 accuracy: {class_0_acc:.2%} ({class_0_correct}/{class_0_total})")
    if class_1_total > 0:
        print(f"  Class 1 accuracy: {class_1_acc:.2%} ({class_1_correct}/{class_1_total})")
        precision = class_1_correct / (class_1_correct + (class_0_total - class_0_correct)) if (class_1_correct + (class_0_total - class_0_correct)) > 0 else 0
        recall = class_1_acc
        f1 = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0
        print(f"  Precision: {precision:.4f}")
        print(f"  Recall: {recall:.4f}")
        print(f"  F1 Score: {f1:.4f}")
    else:
        print(f"  Class 1 accuracy: No class 1 samples found")

    return avg_loss, class_0_acc, class_1_acc

# After all training is complete, reload both models for full evaluation
print("\n" + "="*50)
print("FULL TEST SET EVALUATION")
print("="*50)

# Reload model with PE
model_with_pe, _ = setup_model_with_pe_and_lora(use_pe=True)
model_with_pe = model_with_pe.to(device)
model_with_pe.load_state_dict(torch.load('model_with_pe.pth'))

# Reload model without PE
model_without_pe, _ = setup_model_with_pe_and_lora(use_pe=False)
model_without_pe = model_without_pe.to(device)
model_without_pe.load_state_dict(torch.load('model_without_pe.pth'))

# Run full evaluations
print("\n" + "="*60)
print("COMPLETE TEST SET EVALUATION - MODEL WITH PE")
print("="*60)
test_results_with_pe = evaluate_with_progress(model_with_pe, test_loader, "Test Set WITH PE", print_every=100)

print("\n" + "="*60)
print("COMPLETE TEST SET EVALUATION - MODEL WITHOUT PE")
print("="*60)
test_results_without_pe = evaluate_with_progress(model_without_pe, test_loader, "Test Set WITHOUT PE", print_every=100)





FULL TEST SET EVALUATION
Loading EleutherAI/pythia-160m...
Using positional encodings: True
Total parameters: 162,862,848
Trainable parameters: 539,904
Trainable %: 0.33%
Loading EleutherAI/pythia-160m...
Using positional encodings: False
Total parameters: 162,839,040
Trainable parameters: 516,096
Trainable %: 0.32%

COMPLETE TEST SET EVALUATION - MODEL WITH PE

Starting Test Set WITH PE evaluation on 823 batches...
  Progress: 100/823 batches (12.2%)
  Current - Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%
  Progress: 200/823 batches (24.3%)
  Current - Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%
  Progress: 300/823 batches (36.5%)
  Current - Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%
  Progress: 400/823 batches (48.6%)
  Current - Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%
  Progress: 500/823 batches (60.8%)
  Current - Loss: 0.0000, Class 0: 100.00%, Class 1: 100.00%


KeyboardInterrupt: 